crates/sui-core/src/consensus_handler.rs

```diff
@@ -703,15 +713,31 @@ impl<C: CheckpointServiceNotify + Send + Sync> ConsensusHandler<C> {
         {
             let span = trace_span!("ConsensusHandler::HandleCommit::process_consensus_txns");
             let _guard = span.enter();
-            for (authority_index, parsed_transactions) in consensus_commit.transactions() {
+            for (block_ref, parsed_transactions) in consensus_commit.transactions() {
+                let author = block_ref.author.value();
                 // TODO: consider only messages within 1~3 rounds of the leader?
-                self.last_consensus_stats
-                    .stats
-                    .inc_num_messages(authority_index as usize);
-                for parsed in parsed_transactions {
+                self.last_consensus_stats.stats.inc_num_messages(author);
+                for (tx_index, parsed) in parsed_transactions.into_iter().enumerate() {
                     // Skip executing rejected transactions. Unlocking is the responsibility of the
                     // consensus transaction handler.
                     if parsed.rejected {
+                        if let Some(mysticeti_rejected_transactions) =
+                            self.epoch_store.mysticeti_rejected_transactions.as_ref()
+                        {
+                            if matches!(
+                                parsed.transaction.kind,
+                                ConsensusTransactionKind::UserTransaction(_)
+                            ) {
+                                mysticeti_rejected_transactions.reject_transaction(
+                                    MysticetiTransactionPosition {
+                                        block_ref,
+                                        transaction_index: tx_index as u32,
+                                    },
+                                );
+                            }
+                            // TODO: We also need to eventually revert any transaction
+                            // that was previously fast-path certified but now rejected.
```

Comment:
If we don't lock based on non-finalized fast path execution results than we dont need this revert right?